File: netlogo-headless/src/main/prim/_setobservervariable.scala
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.prim

import org.nlogo.api.AgentException
import org.nlogo.nvm.{ Command, Context }

class _setobservervariable(_vn: Int) extends Command {

  def this(original: _observervariable) = this(original.vn)

  switches = true

  override def toString =
    super.toString + ":" +
      Option(world).map(_.observerOwnsNameAt(vn)).getOrElse(vn.toString)

  // MethodRipper won't let us call a public method from report_1()
  // so we must keep vn and _vn separate - ST 9/22/12
  def vn = _vn

  override def perform(context: Context): Unit = {
    perform_1(context, args(0).report(context))
  }

  def perform_1(context: Context, arg0: AnyRef): Unit = {
    try world.observer.setVariable(_vn, arg0)
    catch { case ex: AgentException =>
      throw new org.nlogo.nvm.RuntimePrimitiveException(
        context, this, ex.getMessage) }
    context.ip = next
  }

}

File: netlogo-gui/src/main/prim/_setobservervariable.java
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.prim;

import org.nlogo.api.LogoException;
import org.nlogo.core.Syntax;
import org.nlogo.nvm.Command;

public final strictfp class _setobservervariable
    extends Command {
  int vn = 0;

  public _setobservervariable(_observervariable original) {
    vn = original.vn;
    this.switches = true;
  }



  @Override
  public String toString() {
    if (world != null) {
      return super.toString() + ":" + world.observerOwnsNameAt(vn);
    } else {
      return super.toString() + ":" + vn;
    }
  }

  @Override
  public void perform(final org.nlogo.nvm.Context context) throws LogoException {
    try {
      world.observer().setVariable(vn, args[0].report(context));
    } catch (org.nlogo.api.AgentException ex) {
      throw new org.nlogo.nvm.RuntimePrimitiveException
          (context, this, ex.getMessage());
    }
    context.ip = next;
  }

  public void perform_1(final org.nlogo.nvm.Context context, Object arg0) throws LogoException {
    try {
      world.observer().setVariable(vn, arg0);
    } catch (org.nlogo.api.AgentException ex) {
      throw new org.nlogo.nvm.RuntimePrimitiveException
          (context, this, ex.getMessage());
    }

    context.ip = next;
  }
}
